src/data-fetchers/vcf/vcf-worker.ts

```diff
@@ -70,80 +143,31 @@ const tilesetInfo = (uid: string) => {
+ * @param z A number which is the z coordinate of the tile
+ * @param x A number which is the x coordinate of the tile
+ */
 const tile = async (uid: string, z: number, x: number): Promise<void[]> => {
     const source = dataSources.get(uid)!;
-    const parser = await source.file.getParser();
-
     const CACHE_KEY = `${uid}.${z}.${x}`;
 
     // TODO: Caching is needed
     // if (!tileValues[CACHE_KEY]) {
     tileValues[CACHE_KEY] = [];
     // }
-
-    const recordPromises: Promise<void>[] = [];
     const tileWidth = +source.tilesetInfo.max_width / 2 ** +z;
 
     // get bounds of this tile
     const minX = source.tilesetInfo.min_pos[0] + x * tileWidth;
     const maxX = source.tilesetInfo.min_pos[0] + (x + 1) * tileWidth;
 
-    let curMinX = minX;
-
-    const { chromLengths, cumPositions } = source.chromInfo;
-
-    cumPositions.forEach(cumPos => {
-        const chromName = cumPos.chr;
-        const chromStart = cumPos.pos;
-        const chromEnd = cumPos.pos + chromLengths[chromName];
-
-        const parseLineStoreData = (line: string, prevPos?: number) => {
-            const vcfRecord: VcfRecord = parser.parseLine(line);
-            const data = recordToTile(vcfRecord, cumPos.pos, prevPos);
-
-            // Store this column
-            tileValues[CACHE_KEY] = tileValues[CACHE_KEY].concat([data]);
-
-            // Return current POS
-            return vcfRecord.POS;
-        };
-
-        let startPos, endPos;
-        if (chromStart <= curMinX && curMinX < chromEnd) {
-            // start of the visible region is within this chromosome
-            let prevPOS: number | undefined;
-            if (maxX > chromEnd) {
-                // the visible region extends beyond the end of this chromosome
-                // fetch from the start until the end of the chromosome
-
-                startPos = curMinX - chromStart;
-                endPos = chromEnd - chromStart;
-                recordPromises.push(
-                    source.file.tbi
-                        .getLines(chromName, startPos, endPos, line => {
-                            prevPOS = parseLineStoreData(line, prevPOS);
-                        })
-                        .then(() => {})
-                );
-            } else {
-                startPos = Math.floor(curMinX - chromStart);
-                endPos = Math.ceil(maxX - chromStart);
-                recordPromises.push(
-                    source.file.tbi
-                        .getLines(chromName, startPos, endPos, line => {
-                            prevPOS = parseLineStoreData(line, prevPOS);
-                        })
-                        .then(() => {})
-                );
-                return;
-            }
-
-            curMinX = chromEnd;
-        }
+    const recordPromises = await source.file.getTileData(minX, maxX, (vcfTile: VcfTile) => {
+        tileValues[CACHE_KEY] = tileValues[CACHE_KEY].concat([vcfTile]);
```

Comment:
I believe `Array.concat` returns a new copy of the merged array, which could be expensive if the number of records gets very large. I think it should be ok to mutate `tileValues[CACHE_KEY].push(vcfTile)`